python/ray/dag/dag_node_operation.py

```diff
@@ -0,0 +1,337 @@
+
+    def __lt__(self, other: "_DAGOperationGraphNode"):
+        """
+        Two _DAGOperationGraphNodes are comparable only when they belong to
+        the same actor. For operations on the same actor, if `local_idx` is smaller,
+        the DAGNode to which this operation belongs has a smaller `bind_index`.
+        """
+        assert self.actor_handle == other.actor_handle
+        return self.operation.local_idx < other.operation.local_idx
+
+    def __eq__(self, other: "_DAGOperationGraphNode"):
+        """
+        Two _DAGOperationGraphNodes are comparable only when they belong to the
+        same actor. For operations on the same actor, two operations are equal
+        only when they have the same `local_idx` and `type`.
+        """
+        assert self.actor_handle == other.actor_handle
+        return (
+            self.operation.local_idx == other.operation.local_idx
+            and self.operation.type == other.operation.type
+        )
+
+    def __hash__(self):
+        """
+        An operation is uniquely identified by its `dag_idx` and type.
+        """
+        return hash((self.operation, self.dag_idx))
+
+
+def _add_edge(from_node: _DAGOperationGraphNode, to_node: _DAGOperationGraphNode):
+    """
+    Add an edge from `from_node` to `to_node`. An edge is a tuple of
+    the operation's `dag_idx` and type.
+    """
+    from_node.out_edges.add((to_node.dag_idx, to_node.operation.type))
+    to_node.in_edges.add((from_node.dag_idx, from_node.operation.type))
+
+
+def _select_next_nodes(
+    actor_to_candidates: Dict["ray._raylet.ActorID", List[_DAGOperationGraphNode]],
+    graph: Dict[int, Dict[_DAGNodeOperationType, _DAGOperationGraphNode]],
+):
+    """
+    This function selects the next nodes for topological sort to generate execution
+    schedule. If there are multiple candidate _DAGOperationGraphNodes, select nodes
+    based on the following rules:
+
+    #1  If the nodes are not NCCL write nodes, select the one with the smallest
+        `bind_index`. If there are multiple candidate nodes with the smallest
+        `bind_index` among the actors to which they belong, any one of them is
+        acceptable, but the implementation ensures the result is deterministic.
+        For the implementation details, we maintain a priority queue for each actor,
+        where the head of the priority queue is the node with the smallest `bind_index`.
+    #2  If #1 cannot be satisfied, it means that all candidate nodes are NCCL write
+        nodes. In this case, select the one at the head of the priority queue and
+        its immediately downstream nodes, which are NCCL read nodes, regardless of
+        whether the downstream nodes are heads of their own priority queues.
+
+    This function may return multiple nodes if they are NCCL nodes. In that case,
+    this function only removes the NCCL write node, which is also the head of a
+    priority queue. Other nodes will be removed in the following iterations.
```

Comment:
Update the comment to say that the READ nodes will be returned even though they should not yet be in the candidate list.

Reply:
Updated https://github.com/ray-project/ray/pull/46911/commits/3241cdaf6bcee89b4563163d8e9ceafd59bc99c8